gitlabform/configuration/transform.py

```diff
@@ -39,82 +41,92 @@ def transform(self, configuration: Configuration) -> None:
         self.user_transformer.transform(configuration)
         self.group_transformer.transform(configuration)
         self.implicit_name_transformer.transform(configuration)
-        self.access_level_transformer.transform(configuration)
+        self.access_level_transformer.transform(configuration, last=True)
 
 
 class ConfigurationTransformer(ABC):
+    def transform(self, configuration: Configuration, last: bool = False) -> None:
+        self._do_transform(configuration)
+        if last:
+            self.convert_to_simple_types(configuration)
+
     @abstractmethod
-    def transform(self, configuration: Configuration) -> None:
+    def _do_transform(self, configuration: Configuration) -> None:
         pass
 
+    @staticmethod
+    def convert_to_simple_types(configuration: Configuration):
+        # we needed complex ruamel.yaml's types like ordereddict and CommentedSeq
+        # for transformations, but at the end convert them to simple dict and lists
+        # for easier to understand debug output and tests
+
+        config_yaml_string = ez_yaml.to_string(obj=configuration.config, options={})
+        simple_yaml_loader = YAML(typ="safe", pure=True)
+        configuration.config = simple_yaml_loader.load(config_yaml_string)
+
 
 class UserTransformer(ConfigurationTransformer):
     def __init__(self, gitlab: GitLab):
         self.gitlab = gitlab
 
-    def transform(self, configuration: Configuration) -> None:
+    def _do_transform(self, configuration: Configuration) -> None:
         logging_args = SimpleNamespace(quiet=False, verbose=False, debug=False)
+        log = ConsolePrinter(logging_args)
+        processor = Processor(log, configuration.config)
 
-        processor = Processor(ConsolePrinter(logging_args), configuration.config)
-
-        paths_to_user = [
-            "projects_and_groups.*.protected_environments.*.deploy_access_levels.user"
-        ]
-
-        for path in paths_to_user:
-            try:
-                for node_coordinate in processor.get_nodes(path):
-                    user = node_coordinate.parent.pop("user")
-
-                    node_coordinate.parent["user_id"] = self.gitlab._get_user_id(user)
-            except YAMLPathException as e:
-                # this just means that we haven't found any keys in YAML
-                # under the given path
-                pass
+        try:
+            for node_coordinate in processor.get_nodes(
```

Comment:
When querying data -- especially when using wildcards -- be sure to set `mustexist=True` as shown in [the YAML Path documentation](https://github.com/wwkimball/yamlpath/wiki/Python-Library#getting-data-from-the-document) which illustrates how to get data out of the document.  Leaving this flag at the default will cause the engine to attempt to generate any missing path ancestry, which is unknown where wildcards are used.